pkg/virtual/framework/wrappers/rbac/merging.go

```diff
@@ -240,40 +241,53 @@ func (l mergedRoleBindingNamespaceLister) Get(name string) (*rbacv1.RoleBinding,
 	panic("implement me")
 }
 
-// TODO: to be converted to generics once we rebase on Go 1.18.
-func mergeRoles(array []*rbacv1.Role, rolesToBeMergedIn []*rbacv1.Role) []*rbacv1.Role {
-	for _, roleToBeMerged := range rolesToBeMergedIn {
-		foundIndex := -1
-		for i, role := range array {
-			if roleToBeMerged.Name == role.Name {
-				foundIndex = i
-				break
+func copyAndMergeRoles(current []*rbacv1.Role, cachedRoles []*rbacv1.Role) []*rbacv1.Role {
+newroles:
+	for _, cachedRole := range cachedRoles {
+		// Have we already seen this Role?
+		for i, existing := range current {
+			if existing.Name == cachedRole.Name {
+				current[i].Rules = mergeRules(current[i].Rules, cachedRole.Rules)
+				continue newroles
 			}
 		}
-		if foundIndex != -1 {
-			array[foundIndex].Rules = append(array[foundIndex].Rules, roleToBeMerged.Rules...)
-		} else {
-			array = append(array, roleToBeMerged)
-		}
+
+		// The Role is a new one, deepcopy it in the current array.
+		current = append(current, cachedRole.DeepCopy())
```

Comment:
use DeepCopyInto. It saves allocations.

Reply:
Why don't we keep track of deepcopied roles and only do it when the second instance is seen?